include/swift/SILOptimizer/Utils/ConstExpr.h

```diff
@@ -40,6 +42,89 @@ class SymbolicValueAllocator;
 class ConstExprFunctionState;
 class UnknownReason;
 
+class StringLiteralInfo {
+  StringLiteralInfo() = default;
+
+public:
+  bool isAscii;
+  StringRef value;
+
+  static Optional<StringLiteralInfo> create(SILValue v) {
+    if (auto *inst = dyn_cast<ApplyInst>(v))
+      return create(inst);
+    return {};
+  }
+
+  static Optional<StringLiteralInfo> create(SILInstruction *inst) {
+    ApplyInst *makeStr = getStringMakeUTF8Apply(inst);
+    if (!makeStr)
+      return {};
+
+    auto strVal = getUTF8String(makeStr);
+    if (!makeStr)
+      return {};
+
+    StringLiteralInfo info;
+    info.value = strVal.getValue();
+    info.isAscii = getIsAscii(makeStr);
+    return info;
+  }
+
+  /// If the given instruction is a call to the compiler-intrinsic initializer
+  /// of String that accepts string literals, return the called function.
+  /// Otherwise, return nullptr.
+  static SILFunction *getStringMakeUTF8Init(SILInstruction *inst) {
+    auto *apply = dyn_cast<ApplyInst>(inst);
+    if (!apply)
+      return nullptr;
+
+    SILFunction *callee = apply->getCalleeFunction();
+    if (!callee || !callee->hasSemanticsAttr("string.makeUTF8"))
+      return nullptr;
+    return callee;
+  }
+
+  /// Similar to getStringMakeUTF8Init but, gets the apply instruction instead.
+  static ApplyInst *getStringMakeUTF8Apply(SILInstruction *inst) {
+    auto *apply = dyn_cast<ApplyInst>(inst);
+    if (!apply)
+      return nullptr;
+
+    SILFunction *callee = apply->getCalleeFunction();
```

Comment:
Couldn't getStringMakeUTF8Init be implemented in terms of this?

Reply:
Yes, but then we would have to call `getCalleeFunction` twice.